pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java

```diff
@@ -0,0 +1,353 @@
+
+    // Delete the files and directories
+    try {
+      FileSystems.delete(fileResources);
+      FileSystems.delete(directoryResourceIdList);
+    } catch (DirectoryNotEmptyException e) {
+      logger.warn(
+          "Error while deleting a directory, ignoring this and resuming further. error={}",
+          e.getMessage());
+    }
+  }
+
+  private TreeSet<String> getRecentSnapshots(List<ResourceId> paths, int numberOfSnapshotsToReturn)
+      throws IOException {
+    TreeSet<String> treeSet = new TreeSet<>();
+    for (ResourceId resourceId : paths) {
+      // Ignore incomplete DWH snapshots
+      if (!isDWHComplete(resourceId)) {
+        continue;
+      }
+      // Keep only the recent snapshots
+      if (treeSet.size() < numberOfSnapshotsToReturn) {
+        treeSet.add(resourceId.getFilename());
+      } else if (resourceId.getFilename().compareTo(treeSet.first()) > 0) {
+        treeSet.add(resourceId.getFilename());
+        treeSet.pollFirst();
+      }
+    }
+    return treeSet;
+  }
+
+  /**
+   * This method tells us whether the DWH snapshot has been successfully completed by the pipeline.
+   * This is determined by checking the existence of both the timestamp files which needs to be
+   * created at the start and end of the pipeline runs.
+   *
+   * @param dwhResource The dwh resource path which needs to be checked for completeness
+   * @return the status of DWH completeness.
+   * @throws IOException
+   */
+  public boolean isDWHComplete(ResourceId dwhResource) throws IOException {
+    ResourceId startTimestampResource =
+        dwhResource.resolve(DwhFiles.TIMESTAMP_FILE_START, StandardResolveOptions.RESOLVE_FILE);
+    ResourceId endTimestampResource =
+        dwhResource.resolve(DwhFiles.TIMESTAMP_FILE_END, StandardResolveOptions.RESOLVE_FILE);
+    if (doesFileExist(startTimestampResource) && doesFileExist(endTimestampResource)) {
+      return true;
+    }
+    return false;
+  }
+
+  private boolean doesFileExist(ResourceId resourceId) throws IOException {
+    List<MatchResult> matchResultList =
+        FileSystems.matchResources(Collections.singletonList(resourceId));
+    MatchResult matchResult = Iterables.getOnlyElement(matchResultList);
+    if (matchResult.status() == Status.OK) {
+      return true;
+    } else if (matchResult.status() == Status.NOT_FOUND) {
+      return false;
+    } else {
+      String errorMessage =
+          String.format("Error matching file spec %s: status %s", resourceId, matchResult.status());
+      logger.error(errorMessage);
+      throw new IOException(errorMessage);
+    }
+  }
+
+  /**
+   * This method returns the base directory where the DWH snapshots needs to be created. This is
+   * determined by ignoring the prefix part in the given input format <baseDir></prefix> for the
+   * dwhRootPrefix
+   *
+   * @param dwhRootPrefix
+   * @return the base directory name
+   */
+  public String getBaseDir(String dwhRootPrefix) {
```

Comment:
I think for this and some of the other methods in this class, we don't really need `public` and package-private should be enough; if yes, please drop `public` where it is not needed.

Reply:
This method is used by the PipelineManager.java as well, hence this is public. Also, made sure only the methods which are externally used are made public.

Reply:
Right but those are all in the same *package* i.e., `org.openmrs.analytics`, right? So I think package-private should still work. Not a big deal but I usually use "public" when I expect usage from outside my package.

Reply:
Agreed, have made the methods as package-private.